engine/src/snapshot/logic.rs

```diff
@@ -59,120 +63,105 @@ pub enum WriteError {
-    // create ephemeral key pair.
-    let ephemeral_key = x25519::SecretKey::generate().map_err(|e| WriteError::GenerateRandom(format!("{}", e)))?;
-
-    // get public key.
-    let ephemeral_pk = ephemeral_key.public_key();
 
-    let ephemeral_pk_bytes = ephemeral_pk.to_bytes();
+    #[error("unsupported associated data")]
+    UnsupportedAssociatedData,
 
-    // write public key into output.
-    output.write_all(&ephemeral_pk_bytes)?;
-
-    // secret key now expects an array
-    let mut key_bytes = [0u8; x25519::SECRET_KEY_LENGTH];
-    key_bytes.clone_from_slice(key);
-
-    // get `x25519` secret key from public key.
-    let pk = x25519::SecretKey::from_bytes(key_bytes).public_key();
-
-    let pk_bytes = pk.to_bytes();
-
-    // do a diffie_hellman exchange to make a shared secret key.
-    let shared = ephemeral_key.diffie_hellman(&pk);
-
-    // compute the nonce using the ephemeral keys.
-    let nonce = {
-        let mut i = ephemeral_pk.to_bytes().to_vec();
-        i.extend_from_slice(&pk_bytes);
-        let res = blake2b::Blake2b256::digest(&i).to_vec();
-        let v: Nonce = res[0..NONCE_SIZE].try_into().expect("slice with incorrect length");
-        v
-    };
-
-    // create the XChaCha20Poly1305 tag.
-    let mut tag = [0; XChaCha20Poly1305::TAG_LENGTH];
-
-    // creates the ciphertext.
-    let mut ct = vec![0; plain.len()];
-
-    // decrypt the plain text into the ciphertext buffer.
-    XChaCha20Poly1305::try_encrypt(&shared.to_bytes(), &nonce, associated_data, plain, &mut ct, &mut tag)
-        .map_err(|e| WriteError::CorruptedData(format!("Encryption failed: {}", e)))?;
+    #[error("incorrect work factor")]
+    IncorrectWorkFactor,
+}
 
-    // write tag and ciphertext into the output.
-    output.write_all(&tag)?;
-    output.write_all(&ct)?;
+/// Encrypt snapshot content with key using work factor recommended for password-based (weak) keys.
+///
+/// # Security
+///
+/// Weak low-entropy (password-based) keys must be strengthened with key derivation.
+/// Secure key derivation in this case is resource-consuming, ie. it can use a lot of RAM
+/// and should take approx. 1 second. Strong keys generated with cryptographically secure RNG
+/// don't need strengthening.
+///
+/// This function expects key to be password-based (blake2b hash of the user provided password).
+/// It uses recommended work factor (approx. 20) to derive encryption key.
+/// It is safe to use with strong keys, although computing resources may be wasted.
+/// In this case it is recommended to use `encrypt_content_with_work_factor` with small/zero work factor.
+pub fn encrypt_content<O: Write>(plain: &[u8], output: &mut O, key: &Key) -> Result<(), WriteError> {
+    let work_factor = age::RECOMMENDED_MINIMUM_ENCRYPT_WORK_FACTOR;
+    // TODO: work_factor is intentionally 0 just for development.
+    // Use proper value in production.
+    // let work_factor = 1;
```

Comment:
Tests (in client) now work slowly because of recommended work factor in the encryption function (in engine). I use this comment to change work factor so that tests work faster. Is it possible to do that automatically? `#[cfg(test)]` doesn't work because it is a different crate (tests run in client, but the slow code is in engine).

I can simply remove the "TODO:" part of the comment.

Reply:
I don't think it's possible to do what you want to do without making this public and allow other crates to overwrite it.

Yeah, maybe just remove the `TODO` then.